cep-20.2.md

```diff
@@ -0,0 +1,711 @@
+folder: path
+# absolute or relative path from recipe file
+patches: [path]
+```
+
+### Git source
+
+```yaml
+# URL to the git repository or path to local git repository
+git_url: url | path
+# revision to checkout to (commit or tag)
+git_rev: string
+git_depth: signed integer (defaults to -1 -> not shallow)
+```
+
+### Removed source definitions
+
+SVN and HG (mercury) source definitions are removed as they are not relevant anymore.
+
+## Requirements section
+
+```yaml
+requirements:
+  # build time dependencies, in the build_platform architecture
+  build: [MatchSpec]
+  # dependencies to link against, in the target_platform architecture
+  host: [MatchSpec]
+  # the section below is copied into the index.json and required at package installation
+  run: [MatchSpec]
+  # constrain optional packages
+  run_constrained: [MatchSpec]
+```
+
+## Test section
+
+<details>
+<summary>
+The current state of the Test section which is being removed with this spec.
+
+Note: the test section also has a weird implicit behavior with `run_test.sh`, `run_test.bat` as well as `run_test.py` and `run_test.pl` script files
+that are run as part of the test.
+
+</summary>
+The current YAML format for the test section is:
+
+```yaml
+test:
+  # files (from recipe directory) to include with the tests
+  files: [glob]
+  # files (from the work directory) to include with the tests
+  source_files: [glob]
+  # requirements at test time, in the target_platform architecture
+  requires: [MatchSpec]
+  # commands to execute
+  commands: [string]
+  # imports to execute with python (e.g. `import <string>`)
+  imports: [string]
+  # downstream packages that should be tested against this package
+  downstreams: [MatchSpec]
+```
+</details>
+
+```
+test: [TestElement]
+```
+
+The new test section consists of a list of test elements. Each element is executed independently and can have different requirements. There are multiple types of test elements defined, such as the `command` test element, the `python` test element and the `downstream` test element.
+
+#### Command test element
+
+```yaml
+# script to execute
+script: string | [string]
+# optional extra requirements
+extra_requirements:
+  # extra requirements with build_platform architecture (emulators, ...)
+  build: [MatchSpec]
+  # extra run dependencies
+  run: [MatchSpec]
+# extra files to add to the package for the test
+# Alternatively, use ${{ RECIPE_DIR }} or ${{ SOURCE_DIR }} and a single list?
+# or $RECIPE_DIR / $SRC_DIR ...
+files:
+  # files from $SRC_DIR
+  source: [glob]
+  # files from $RECIPE_DIR
+  recipe: [glob]
+```
+
+#### Python test element
+
+```yaml
+python:
+  # list of imports to try
+  imports: [string]
+```
+
+#### Downstream test element
+
+```yaml
+downstream: MatchSpec
+```
+
+## Outputs section
+
+conda-build has very hard to understand behavior with multiple outputs. We propose some drastic simplifications.
+Each output in the new format has the same keys as a "top-level" recipe.
```

Comment:
I believe that the reason people find multiple-output workflows with conda-build confusing is that it was originally intended to only be used for a "build once, then split" workflow. In that workflow, the outputs [only do one of two things](https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#specifying-files-to-include-in-output): installation to / removal from the $PREFIX via script, or splitting of files already in the $PREFIX via a glob list. However, people have used the `outputs/script` key for scripts that do MORE than just moving files, but also doing additional build steps. i.e. They try to use the top-level recipe as a cached build which is continued and reused multiple times.

My understanding is that this proposal moves the recipe spec away from "build once, then split" towards a graph of multiple build caches and multiple outputs. The only use case for multi-cache multi-output I know of is a mixed C/Python package where a C library is wrapped in Python extensions which must be recompiled for each Python version. By moving to a recipe specification that explicitly has the concept of cached builds, this makes this type of combined multi-output recipe more efficient because the C library is reused? However, for conda-forge at least, feedstocks run each Python build in parallel, so enabling this type of caching will not be immediately beneficial. In practice, I actually encourage maintainers to use separate recipes in the case described above because it avoids multi-output recipes and migrations of the C library don't trigger unnecessary rebuilds of the python package.

I think the proposed "multi-cache, multi-output" workflow is MORE complicated instead of simpler. An actual simplification of the outputs section would remove the `script` option from outputs and only allow pattern matching (but use REGEX to make it more powerful?). This would enforce a "build once, then split" workflow and avoid the common issues which arise when people try to use the output script for doing additional build steps instead of just moving/selecting files.

Perhaps you could be more specific in this proposal about the problem using the current outputs section, some example use cases where it fails, and the type of workflow you want to enable with the new outputs section.

Reply:
You are right, but you are also missing some of the tricky behavior from conda. Namely that requirements from the top level are mixed into the requirements of the output (that has the same name as the top-level recipe).

Splitting packages is also pretty difficult.
I think people are relatively used to re-use caches e.g. from Github actions. I hope we have an implementation of this soon in rattler-build so you can try it out.

Rebuilding the (existing) C library (as you describe) can hopefully also be avoided if we do some more clever things regarding hashes and comparing if builds already exist.

Reply:
On further reflection, I think that these proposed changes to the outputs section are more than just new keys and values. They are actually implying a fundamental change of the documented (canonical?) conda-build workflow (as opposed to just moving or change the names of exisitng keys) and should be their own CEP.

P.S. I think the flatpak build system is designed to have cached modular components with arbitrary dependency graphs, so maybe we can steal some design or implementation ideas from them?

Reply:
I am not opposed to split out the outputs work into another CEP and leave it undefined in this CEP.
For now we'd still implement as laid out here, but open to changes.

Reply:
> mixed C/Python package where a C library is wrapped in Python extensions which must be recompiled for each Python version

This was one of the main workflows intended. Splitting files up was definitely a major use case (for GCC packages back at this feature's inception), but making wrapper builds in a matrix was equally important (for xgboost back at this feature's inception).

Not that history really matters here. What matters is how hard it is to use, and whether it is saving anyone any time.